### PR TITLE
first_even_numbers_sum-my_reverse

### PR DESCRIPTION
**Pull request by cbaker-ut**
- Add description and examples to problems without directions
- Update spec files to include matching descriptions
- Update describe block in spec files to clarify whether a method is monkey-patching a class or free standing, as in `Array#my_all` vs `my_all`
- Backtick all code in method descriptions
- Update specs to move check for blacklisted methods from example to a before hook

### CODE REVIEW

**Comment by cbaker-ut** [comment]
Now added: 
- before hooks blacklisting #each in the `Hash#my_each`, `Array#my_each`
- before hook blacklisting `Array#sort` and `Array#sort!` in `Array#merge_sort`

File: problems/jumble_sort.rb [hunk 1]
```diff
@@ -3,8 +3,8 @@
 # no alphabet is passed in, it defaults to normal alphabetical order (a-z).
 
 # Example:
-# jumble_sort("hello") => "ehllo"
-# jumble_sort("hello", ['o', 'l', 'h', 'e']) => 'ollhe'
+# `jumble_sort("hello")` => "ehllo"
+# `jumble_sort("hello", ['o', 'l', 'h', 'e'])` => 'ollhe'
```
Comment (ronilbhatia) [inline problems/jumble_sort.rb:8]: Let's blacklist `sort` and `sort_by` for this method (dangerous and non-dangerous). Additionally for all methods we blacklist we need to explicitly state as such in the problem description. So add a `before(:each)` hook to test that these methods aren't called on *any instance* of an Array (`expect_any_instance_of(Array)`) and a note to this problem description as follows:

"\*\*Do NOT use the built-in `Array#sort` or `Array#sort_by` methods in your implementation\*\*"

File: problems/my_all.rb [hunk 2]
```diff
@@ -1,3 +1,7 @@
+# Write an `Array#my_all?(&prc)` method. 
+# This method should return true if every element in the array satisfies the block, otherwise return false.
+# Example: `[1,2,3].my_all? {|n| n.even?}` => false
+# `[2,4,6].my_all? {|n| n.even?}` => true
```
Comment (ronilbhatia) [inline problems/my_all.rb]: Add note similar to above explicitly saying they can't use built-in `Array#all?` in their implementation

File: problems/my_any.rb [hunk 3]
```diff
@@ -1,3 +1,8 @@
+# Write an `Array#my_any?(&prc)` method.
+# This method should return true if any of the Array elements satisfy the block,
+# otherwise it should return false.
+# Example: `[1,2,3].my_any? {|n| n.even?}` => true
+# `[1,3,5].my_any? {|n| n.even?}` => false
```
Comment (ronilbhatia) [inline problems/my_any.rb]: Add note similar to above explicitly saying they can't use built-in `Array#any?` in their implementation

File: problems/my_each.rb [hunk 4]
```diff
@@ -1,3 +1,5 @@
+# Write an `Array#my_each(&prc)` method that calls a proc on each element of the array.
+# Write an `Array#my_each_with_index(&prc)` method that calls a proc on each element with its index.
```
Comment (ronilbhatia) [inline problems/my_each.rb]: Add note similar to above explicitly saying they can't use built-in `Array#each`, `Array#each_with_index` or `Array#map` in their implementation

File: problems/my_each_hash.rb [hunk 5]
```diff
@@ -1,5 +1,5 @@
 class Hash
-  # Write a version of my each that calls a proc on each key, value pair
+  # Write a `Hash#my_each(&prc)` that calls a proc on each key, value pair
```
Comment (ronilbhatia) [inline problems/my_each_hash.rb:2]: Add note similar to above explicitly saying they can't use built-in `Hash#each`, `Hash#each_with_index` or `Hash#map` in their implementation

File: problems/my_flatten.rb [hunk 6]
```diff
@@ -1,16 +1,16 @@
...
   end
 
-  # Write a version of flatten that only flattens n levels of an array.
+  # Write an `Array#my_controlled_flatten(n)` method that only flattens n levels of an array.
   # E.g. If you have an array with 3 levels of nested arrays, and run
   # my_flatten(1), you should return an array with 2 levels of nested
   # arrays
   #
-  # [1,[2,3], [4,[5]]].my_controlled_flatten(1) => [1,2,3,4,[5]]
+  # `[1,[2,3], [4,[5]]].my_controlled_flatten(1)` => [1,2,3,4,[5]]
```
Comment (ronilbhatia) [inline problems/my_flatten.rb:13]: Can we separate out `flatten` and `controlled_flatten` into two separate problems? We'll need to do this for the problems, specs, and solutions, and make sure to update `list.csv` with another line for `controlled_flatten` - i would suggest keeping these files for the regular flatten, and making separate ones for `my_controlled_flatten`. Let me know if you're confused about this and we can discuss.

File: problems/my_each.rb (hunk 4 above)
Comment (ronilbhatia) [reply problems/my_each.rb]: Additionally, I'd like to separate out `each` and `each_with_index` into separate problems, so keep these files for the regular `my_each` and make separate problem, spec, and solution files for `each_with_index`. Let's also change the specs in here so they match what is in our assessment-generator repo, 1 spec is not enough for these problems.

Comment (ronilbhatia) [reply problems/my_each.rb]: One more thing on this - so for the `my_each` problems we want to disallow both `each` and `each_with_index` as well as `map`. However, unlike other problems weird stuff will happen with rspec if we say `expect_any_instance_of(Array).not_to receive(:each)` so here's the workaround I came up with and implemented for A01:

```
let(:arr) { [1,2,3] }
  before(:each) do
    expect(arr).not_to receive(:each)
    expect(arr).not_to receive(:each_with_index)
    expect(arr).not_to receive(:map)
    expect(arr).not_to receive(:map!)
    expect(arr).not_to receive(:dup)
    arr.my_each { |el| el }
  end
```

The idea is we expect the instance itself not to receive any of these methods - but students could easily get around this by duping the array, so we disallow duping of the instance as well. It's not perfect but I think it's good enough and about all we can do. Unless you think of something else, then let me know!

File: problems/my_inject.rb [hunk 7]
```diff
@@ -1,5 +1,5 @@
 class Array
-  # Monkey patch the Array class and add a my_inject method. If my_inject receives
+  # Write an `Array#my_inject` method. If my_inject receives
```
Comment (ronilbhatia) [inline problems/my_inject.rb:2]: Add note similar to above explicitly stating that they are not allowed to use built-in `Array#inject` or `Array#reduce` methods.

File: problems/my_join.rb [hunk 8]
```diff
@@ -1,8 +1,8 @@
-# Monkey patch the Array class and add a my_join method. If my_join receives
+# Write an `Array#my_join` method. If my_join receives
 # no argument, then use an empty string to join the array together.
 # ex.
-# [1,2,3].my_join => '123'
-# [1,2,3].my_join('$') => '1$2$3'
+# `[1,2,3].my_join` => '123'
+# `[1,2,3].my_join('$')` => '1$2$3'
```
Comment (ronilbhatia) [inline problems/my_join.rb]: Add note similar to above explicitly stating that they are not allowed to use built-in `Array#join` method.

File: problems/my_merge.rb [hunk 9]
```diff
@@ -1,5 +1,5 @@
 class Hash
-  # Write a version of merge. This should NOT modify the original hash and 
+  # Write a `Hash#my_merge(other_hash)` method. This should NOT modify the original hash and
```
Comment (ronilbhatia) [inline problems/my_merge.rb:2]: Add note similar to above explicitly stating that they are not allowed to use built-in `Hash#merge` method.

File: problems/my_reject.rb [hunk 10]
```diff
@@ -1,3 +1,6 @@
+# Write an `Array#my_reject(&prc)` method.
+# This method should return a new array excluding all the elements in the original array that satisfy the proc.
+# Example: `[1,2,3].my_reject {|n| n.even?}` => [1,3]
```
Comment (ronilbhatia) [inline problems/my_reject.rb]: So for this one I want to disallow both `select` and `reject`. However, unlike some of the others ones weird stuff will happen with rspec if we say `expect_any_instance_of(Array).not_to receive(:select)` so here's the workaround I came up with and implemented for A01:

```
let(:arr) { [1,2,3] }
  before(:each) do
    expect(arr).not_to receive(:select)
    expect(arr).not_to receive(:dup)
    expect_any_instance_of(Array).not_to receive(:select!)
    expect_any_instance_of(Array).not_to receive(:reject)
    expect_any_instance_of(Array).not_to receive(:reject!)
    arr.my_select { |el| el }
  end
```

The idea is we expect the instance itself not to receive `select` - but students could easily get around this by duping the array, so we disallow the `dup` method as well on that instance. It's not perfect but I think it's good enough and about all we can do. Unless you think of something else, then let me know!

File: problems/merge_sort.rb [hunk 11]
```diff
@@ -1,5 +1,5 @@
+# Write an `Array#merge_sort` method; it should not modify the original array.
```
Comment (ronilbhatia) [inline problems/merge_sort.rb]: Let's blacklist `sort` and `sort_by` for this method (dangerous and non-dangerous). Additionally for all methods we blacklist we need to explicitly state as such in the problem description. So add a `before(:each)` hook to test that these methods aren't called on *any instance* of an Array (`expect_any_instance_of(Array)`) and a note to this problem description as follows:

"\*\*Do NOT use the built-in `Array#sort` or `Array#sort_by` methods in your implementation\*\*"

File: problems/my_each_hash.rb (hunk 5 above)
Comment (ronilbhatia) [reply problems/my_each_hash.rb:2]: Let's also add more specs to this one - you can essentially follow the format for what we do in the assessment generator for `Array#my_each` but obviously updating it to apply to a hash. Might be a bit tricky, if you need help or are having trouble figuring out how to do it lmk and I can take a look after you've updated for everything else.

**Review by ronilbhatia — CHANGES_REQUESTED** [review]
Hey Connor - great job going through all of this! The majority of the comments that I left are for the following 2 things:

1) Add explicit lines to the problem descriptions stating which methods they are not allowed to use. If there is a dangerous version method of the same name, then you don't need to state that in addition to the non-dangerous version in the description (should be implied), but make sure to test for both of them in the spec file.
2) For blacklist methods in the spec files, replace `expect(instance)` with `expect_any_instance_of(Class)`. There are some exceptions to this that I've detailed in comments on specific files (the `my_each` ones and `my_reject`).
3) Right now there aren't enough specs for `my_each` problems - take a look at what we do in the assessment generator and update the specs to match that. Do this for `Array#my_each`, `Array#my_each_with_index` and `Hash#my_each`
4) Separate `Array#my_each` and `Array#my_each_with_index` into separate files, and update `list.csv` to include both of them. Do the same with `Array#flatten` and `Array#my_controlled_flatten`. Let me know if you have questions on how to update the `list.csv` file properly.

**UPDATE**:

One other thing - the `my_controlled_flatten` solution is wrong - it doesn't consider the `nil` case - can you update it to this instead:

```ruby
  def my_controlled_flatten(level = nil)
    flattened = []

    self.each do |ele|
      if ele.is_a?(Array) && level != 0
        flattened += (level.nil? ? ele.my_controlled_flatten : ele.my_controlled_flatten(level - 1))
      else
        flattened << ele
      end
    end

    flattened
  end
```

File: problems/median.rb [hunk 12]
```diff
@@ -1,4 +1,4 @@
-# Write a method that returns the median of elements in an array
+# Write an `Array#median` method that returns the median element in an array
```
Comment (jliversi) [inline problems/median.rb]: Move this prompt into the class statement, this is how it's done on other problems and the real assessment and we should be consistent. ex.
```
class Array
     # prompt here
     def method_name
     end
end
```

File: problems/merge_sort.rb (hunk 11 above)
Comment (jliversi) [inline problems/merge_sort.rb]: Move this prompt into the class statement

File: problems/my_all.rb [hunk 13]
```diff
@@ -1,3 +1,7 @@
+# Write an `Array#my_all?(&prc)` method.
```
Comment (jliversi) [inline problems/my_all.rb]: Move this prompt into the class statement

File: problems/my_any.rb [hunk 14]
```diff
@@ -1,3 +1,8 @@
+# Write an `Array#my_any?(&prc)` method.
```
Comment (jliversi) [inline problems/my_any.rb]: Move this prompt into the class statement

File: problems/my_each.rb [hunk 15]
```diff
@@ -1,3 +1,5 @@
+# Write an `Array#my_each(&prc)` method that calls a proc on each element of the array.
```
Comment (jliversi) [inline problems/my_each.rb]: Move these prompts into the class statement

File: problems/my_join.rb [hunk 16]
```diff
@@ -1,8 +1,8 @@
-# Monkey patch the Array class and add a my_join method. If my_join receives
+# Write an `Array#my_join` method. If my_join receives
```
Comment (jliversi) [inline problems/my_join.rb]: Move this prompt into the class statement

File: problems/my_reject.rb [hunk 17]
```diff
@@ -1,3 +1,6 @@
+# Write an `Array#my_reject(&prc)` method.
```
Comment (jliversi) [inline problems/my_reject.rb]: Move this prompt into the class statement

**Review by jliversi — CHANGES_REQUESTED** [review]

**Comment by jliversi** [comment]
Just a few quick changes and then I think it's good to merge:
- my_each_with_index: add empty line at the end of problem

- add one more spec to `my_flatten` and `my_controlled_flatten` (I think it's good to aim for a minimum of two specs per problem). A simple additional "works on an already flattened array" should work for both

- Similar note, add one more spec to my_reverse (maybe just to check that it works on an empty or single-element array) 

- Currently the `my_all` spec breaks things because of `let(:arr) = [1,2,3]`, just change this to a block instead of `=` 

- For `my_any` and `my_inject`, use let to instantiate the array instead of just using a variable. 

- For the following problems, remove the call to the method inside the `before(:each)` block. Because the methods are invoked IN the `it` blocks, the extra call is unnecessary
	- jumble_sort
	- merge_sort
	- my_all
	- my_any
	- my_each_hash
	- my_each 
	- my_each_with_index
	- both my_flattens
	- my_inject 
	- my_join
	- my_merge
	- my_reject
	- my_reverse

**Pull request merged**